doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,715 @@
+     have to be changed to use the new API. Such changes are expected to be
+     necessary in very few project.
+
+* **dtype implementors (C-API)**:
+
+  * The array which is currently provided to some functions (such as cast functions),
+    will no longer be provided.
+    For example ``PyArray_Descr->f->nonzero`` or ``PyArray_Descr->f->copyswapn``,
+    may instead receive a dummy array object with only some fields (mainly the
+    dtype), being valid.
+    At least in some code paths, a similar mechanism is already used.
+
+  * The ``scalarkind`` slot and registration of scalar casting will be
+     removed/ignored without replacement.
+     It currently allows partial value-based casting.
+     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
+     but will not be used internally and be deprecated.
+
+   * Currently user dtypes are defined as instances of ``np.dtype``.
+     The creation works by the user providing a prototype instance.
+     NumPy will need to modify at least the type during registration.
+     This has no effect for either ``rational`` or ``quaternion`` and mutation
+     of the structure seems unlikely after registration.
+
+Since there is a fairly large API surface concerning datatypes, further changes
+or the limitation certain function to currently existing datatypes is
+likely to occur.
+For example functions which use the type number as input
+should be replaced with functions taking DType classes instead.
+Although public, large parts of this C-API seem to be used rarely,
+possibly never, by downstream projects.
+
+
+
+Detailed Description
+--------------------
+
+This section details the design decisions covered by this NEP.
+The subsections correspond to the list of design choices presented
+in the Scope section.
+
+Datatypes as Python Classes (1)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The current NumPy datatypes are not full scale python classes.
+They are instead (prototype) instances of a single ``np.dtype`` class.
+Changing this means that any special handling, e.g. for ``datetime``
+can be moved to the Datetime DType class instead, away from monolithic general
+code (e.g. current ``PyArray_AdjustFlexibleDType``).
+
+The main consequence of this change with respect to the API is that
+special methods move from the dtype instances to methods on the new DType class.
+This is the typical design pattern used in Python.
+Organizing these methods and information in a more Pythonic way provides a
+solid foundation for refining and extending the API in the future.
+The current API cannot be extended due to how it is exposed publically.
+This means for example that the methods currently stored in ``ArrFunctions``
+on each datatype (see NEP 40) will be defined differently in the future and
+deprecated in the long run.
+
+The most prominent visible side effect of this will be that
+``type(np.dtype(np.float64))`` will not be ``np.dtype`` anymore.
```

Comment:
Will it be the case that `type(np.dtype(np.float64)) is type(np.dtype(np.float32))`?

Or phrased another way, is this NEP deliberately leaving that question open?

Reply:
My intent is that is False, I am happy to define that right now. If it was not False, `type(np.dtype(np.float64))` would be parametric and I think it is desirable to simplify life by having non-parametric DTypes were possible. They seems easier to me to reason about, and to define ufunc dispatching for.
For the current buildin dtypes, it also seems only right, since they already have ArrFuncs and typenumbers that are distinct.

Of course in general it should be possible to do the above, although I expect some methods may be less efficient for such datatypes, since they may need to test. It is also something to keep in mind, whether during ufunc dispatching datatype instance specialization should be allowed (i.e. allow such a float32+float64 addition loop). But, possibly this can also be added later. The other thing is that it probably can always be possible to define a float32+float64 abstract DType.

Reply:
It sounds then like you're proposing singleton instances for each of the builtin numeric dtypes? Without arguing about whether that makes sense or not yet, I think if that's really your intent, you should get the word `singleton` in a description somewhere further up.

Reply:
Right, almost singleton... We do have byte swapping (and unfortunately metadata). Eric, if you are unsure that is the correct path, should we chat about that briefly?

Reply:
Sorry, almost, since there is byte swapping, and unfortunately metadata.

Reply:
> Eric, if you are unsure that is the correct path, should we chat about that briefly?

I need to set aside a reasonable block of time to finish my train of thought in that direction before such a chat would be useful, and large blocks of time are seemingly hard to come by.

Reply:
Yeah, that is always the problem (except for me, I got too much time to get lost in it). Maybe just for your consideration, why I am thinking along this line:

1. We currently have the methods (in `PyArray_ArrFuncs` defined for each type number basically) and which are housed very similarly to a class.
2. I am thinking that `isinstance(..., np.dtype[float64])` can be useful in general, so it might as well be its own class.
3. The more specific classes are, the easier it seems to quickly dispatch in the UFunc machinery. If float64 is its own class, a loop registered for `float64, float64 -> float64` does not need any additional checks to find the correct C-function (except possibly byte-swapping if you would want to optimize it/skip the casting).
4. I think a DType class is, especially for parametric dtypes, is the most reasonable basis for the first/main step in ufunc dispatching. It seems desirable that classes are thus as specific as useful in that step. It also makes incremental loop addition easier: you can define a ufunc for float64, and then someone else adds the float32 loop.

Of course there are some interesting thoughts, e.g. we could make most methods work more like `get_nonzerofunc(self)` style (why shouldn't nonzero optimize to skip the byte swap check).
But I suppose my angle is this: *It seems useful to me to make the classes as specific as conveniently possible. In other words fully specific aside from storage information (byte swap, string length) and parameters (string length, units).*
Basically, byte-order being considered inconvenient, otherwise it would be true singletons and the main reason to have instances would be for metadata, and because parametric dtypes need something like instances.

Reply:
I started writing up some thoughts here: https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b.

I think I've convinced myself that dtype subclasses are the right way to go, and that metaclasses are not. There's one TODO left in there, which is to look at what `metaclass.__new__` does to store its state - this should give us some hints about where `metadtype.__new__` ought to store its state, and will hopefully put me on the right train of thought to think about singletons.

Reply:
Thanks, I will have to look through this (right now I am not even sure if you mean dtype as classes or a DTypeMeta to define the storage on the DType metaclass).

Reply:
I'm afraid I wrote that after not having read this for a while, so it's based only off my memory of this NEP, and likely does not agree on terminology.

Reply:
Sure. I have tried to parse it. To be sure, you disagree with the metaclass thought to add something similar to `nb_numbers` slots (i.e. methods)? Or are you saying that things like `itemsize` and `flags` should not be be part of the a DType class (metaclass instance), which I fully agree with? I would argue that the "table of behaviours" *is* normally part of the type/class, while all the other things are *not*.

Reply:
I could see that some functions could be defined/specialized for the actual dtype instance (as used in this proposal, e.g. for byte-order or string length). I do think we have a few `staticmethod`s that we need, e.g. because we need to find the correct string length. Although of course there are other ways to approach that, for example using "flexible" instances as we are doing now (which cannot be attached to an array)...

Reply:
I think what I wrote is more an argument against using the word "metaclass". dtypes are not types, so metadtypes are not metatypes. I've updated the doc with a table of comparisons between builtin scalar types and our array types. I don't think anything I've written disagrees with this NEP yet.

Reply:
Right, `dtypes` are not types, but instances of `DType` classes (types). For the user that is all they should care about.
For us, I do currently think we do need some low level C-storage on the DType classes as well as control over `__new__` so technically, I think the DType class has to be an instance of a `DTypeMeta` class, although if you have another idea I am happy to listen.

Its 3 levels deep, so confusing :(, but on the user side it should only be 2 levels `DType` classes (current type number) and `dtype` instances (current dtypes). Even implementors should not really have to care about it too much, they define a new `DType` class, much like a python class, but with slightly different API.

(Not sure this is actually helping, hmm)

Reply:
> For us, I do currently think we do need some low level C-storage on the DType classes

In my tables, which "C decl" row are you thinking that state would need to live inside? I think that's covered without `DtypeMeta`.

You _would_ need `DtypeMeta` (or `IndexableTypeMeta`) to support `np.dtype[...]` as syntax, but that doesn't need to be in an NEP.

Reply:
Currently mainly:
 * Define some metadata about the type (kind, type_num, ...) (depending what `...` is, it may be here, or in the instance part).
 * fast-paths for "magic" methods (f)

Although the "magic" methods, include some like `common_dtype`, which I think are actually class methods (at least in my current view), because they are associated with a `type_num` more than a dtype. I.e. in my view it is odd that we have multiple instances with identical `f` and `type_num`, so those things should be stored on the class. (which effectively makes the class a MetaClass instance, although no python user has to ever think of it that way).

Reply:
That wasn't a particularly clear question from my end, apologies. I was trying to point you to [the following section](https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b#comparing-scalar-types-and-array-types-in-more-depth), but maybe my question didn't make sense anyway.

> I.e. in my view it is odd that we have multiple instances with identical `f` and `type_num`, so those things should be stored on the class.

Note that this is how regular `types` behave today though. If you have:
```
class Foo:
    def __some_slots__(self):
        pass
class Bar(Foo): pass
class Baz(Foo): pass
```
Then at the C level, `((PyTypeObject*)Bar)->tp_some_slot == ((PyTypeObject*)Baz)->tp_some_slot == ((PyTypeObject*)Foo)->tp_some_slot`, thanks to the behavior of `Type_Ready`. Similarly, for python level method overriding / duplication, `type(Foo).__getattribute__` handles the work

So maybe the missing piece of the puzzle is a class heirarchy like:
```
|---- meta-dtypes -----------------|                 |- dtypes------------------------|
dtype <--issubclass-- integral_dtype <--isinstance-- int8 <--issubdtype-- int8_nonative
```
instead of what I think this NEP describes
```
|---- meta-dtypes --------------------------------------------|               |- dtypes----|
dtype <--issubclass-- integral_dtype <--issubclass-- int8_dtype <--instance-- int8
                                                     int8_dtype <--instance-- int8_nonnative
```
Where the distinction is simply _"Does it declare new hook or attribute names at the C level?"_
   * Yes: meta-dtype
   * No: dtype

Reply:
I elaborated on that thought in https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b#what-this-could-mean-for-dtypes.

Reply:
@eric-wieser not sure I can follow, not on the top of my game today, but still. Yes, types use the same C-level slots, but *instances* do not typically. They are meant to be as minimal as possible (not that it matters for dtypes all that much).
One note that I do see `integral_dtype` as useful, but I do not really want to use inheritance for now, it just seems dangerous. E.g. a user integer should not inherit from it.

Note that in your class hierarchy, `int8` is an instance of `integral_dtype` and not `int8`. That moves the precision information into the instance. While possible I feel there is a disassociation with the current `dtypes.num` (which is vital for ufunc dispatching). There are two instances with the same `dtypes.num` (and `issubdtype`), but they have nothing directly in common.

I guess I see it as:

```python
class DType():  # capitalizing but same as np.dtype
   @classmethod
    def __common_dtype__(cls, other):
        raise NotImplementedError

    def __setitem_ptr__(self, ptr, value):
         raise NotImplementedError

class MyDType(DType):
    def __init__(self, itemsize=4):
        self.itemsize = itemsize

    @classmethod
    def __common_dtype__(cls, other):
        raise NotImplementedError

    def __setitem_ptr__(self, ptr, value):
         # does the current logic of ->f->setitem
```
Now in Python, those dunders will be stored in C-side slots on the class, I am suggestion to do the same thing. That requires a DTypeMeta to extend pythons `type` C-slots. While `itemsize` on the other hand is of course an instance variable stored on in the `DType` instance struct.

I am sure you realize all of this though. In my view `integral_dtype` is the same as a Python ABC, mainly something to define a hierarchy.

I guess my main question would be what about the current `dtype.num`? I would want that to be replaced by the class. `dtype.num` is IMO vital, because it is the basis for UFunc dispatching. Furthermore, I do not actually see much of an advantage of `int` being an instance of an `integral` class, it might work within NumPy, but if someone wants to add another integer I am not sure it helps much. And `int8` and `int64` are different enough to have different type numbers (and not all ufuncs may accept all of them).

My guess is what you are proposing is almost the same thing, but the split between what is an instance and what is a class is done at the `integer` level instead of the `int_of_given_precision` level. The main gain would be that you do not have these almost singleton instances. But I am not sure what the actual gain of that is, and I honestly do not quite see how we can allow someone to cramp their own ArbitraryPrecisionInteger into our `np.integer`, it would have to be a subclass? But it is not really a subclass by Liskov substitution principles.

Reply:
> Yes, types use the same C-level slots, but _instances_ do not typically.

The parallel I'm trying to draw is:
1. _scalar instances_ &lrarr; _array instances_
2. _scalar types_ &lrarr; _array dtypes_
3. meta-types &lrarr; meta-dtypes.

With that parallel, it's 3 which declares the slot names, 2 which provides the values of the slots, and 1 which contains a pointer to the table of slots. This NEP seems to be leaning in the direction of

Reply:
I would prefer to write my proposal here as:

1. scalar instances ⇆ array instance + DType instance
2. scalar types ⇆ DType class (type)
3. Metatypes should not be a large priority for now, but `MetaType(type)` ⇆ `MetaDType(DTypeMeta(type))`

Let me try to discuss a few of the points, although it is getting difficult and I may have to take a step back and redefine the terminology here if we want to keep discussing (or make sure to better understand yours).

> This is a feature. isinstance(int8, integral_dtype) conveys the information "int8 carries extra C-level slots specific to integers", which could include
things like ->unsigned_int_type.

Right, however, like in Python were we rarely actually need MetaTypes, I do not think you actually normally need these additional slots on the C-level initially. I may want to enable this, but it seems not a priority to me (i.e. subclassing DTypeMeta). You can already tag on functions to types on the class (in python) well enough, as `int8.unsigned_integer_type()`. Yes, unless we allow subclassing DTypeMeta you cannot do so (efficiently) in C. I am not sure there is much need for it though, at least initially. First, many of these things can be just as well functions likely. Second, it should be possible to spell your example (maybe slightly less efficient) using `np.astype(np.int8, np.unsignedinteger)`.

So I agree with the sentiment that having only a single DTypeMeta (so to say) could be a limitation that should be waved, I am not quite convinced it has to be a priority right now, that integers have their own C-level slots.
To be clear, in my design, I *will* need these things to support things around value-based-casting. I am not opposed to making an `IntegerDTypeMeta` class, I just do not see it as important at this time, and would rather keep that API full private until someone comes along and actually needs it.

> I'd appreciate if you could elaborate on this point, being specific about the types you're referring to. I don't understand it yet.

Sorry, what I meant is that it seems to me for you `>int8`, `<int8`, *and* `int64` are all instances of the same `integer` class. Thus `>int8` and `<int8` must use their `dtype.num` slot to distinguish themselves from `int64`. Alternatively, the UFunc machinery, in a first step, would have to handle all integers identically. But UFuncs are currently designed around `dtype.num` slot. If `>int8` and `<int8` are instances of an `Int8` class (which *could* be an instance of an `IntegerDTypeMeta` if we have a good use for it in the future), then `>int8.type_num` can be replaced by `type(>int8)`, you do not need to store `dtype.num` on the dtype instance.

> It's probably worth comparing to how you'd build an ABC for a regular scalar type in python with C-level slots. My integral_dtype is the the new meta-dtype you
need to support the new slot definition, closer to ABCMeta.

Right (repeating the above), that is all good, it would be fine to have an `IntegerDTypeMeta` (in my terminology), I am just not convinced right now that it is necessary as a first step, especially as public API. I am not opposed though.

> Dispatching rules could use issubdtype(passed_type, dispatch_table_dtype), assuming we could make that fast enough.

Right, I personally feel using `type(dtype)` is a very elegant and easy to understand way though. I assume most speed will have to be fixed by caching anyway, but a `type(dtype)` lookup, is just an `id(type(dtype))` lookup, which should be easy enough to cache well.

> It's done at the "do I need new C attributes and function pointers" level.

Right, and I think a single `DTypeMeta` should be enough for that right now, if it is not, I am happy to start subclassing that `DTypeMeta`. Most Python scalars are also happy with a single metaclass (`type`).

> This sounds like me just choosing the wrong name, integer should have been fixed_integer

Fair enough, although you may want an `integral` as well, which I suppose also in your view could just be more like ABC though.

---

Let me be clear, I am not yet quite clear that we actually have different views here, except where to draw the lines.
I mainly think that a single (public) `DTypeMeta` is enough for now. I think users, and dtype implementors, should (normally) not have to worry about that `DTypeMeta` unless they are very advanced (very few Python programmers need MetaClasses), they just write a few dunders and let Python do the magic in `PyType_Ready`. Could more specialized `DTypeMeta` make sense? Yes, and we probably will need them, but I do not see them as a central concept.

Probably I am confused, but in your code snippet, I see `base_dtype == np.dtype` (sure it _could_ be something else, but I do not really see the need and am wary of implications). In your example `DType == np.dtype` seems wrong, since `DType` is a metaclass so `DType()` should be a class, but it currently is an instance (with an itemsize). Once you do that shift, and assume one `base_dtype == np.dtype` is fine, I think we may be saying very similar things?!

Reply:
> in your code snippet, I see `base_dtype == np.dtype`

I don't see what you mean by that. Can you be more precise? (just edit your comment and I'll delete this comment)

Reply:
Hmm, I guess I mean that `base_dtype == np.dtype` is how I would read it, but I am not sure that you think that would be an OK thing to do (in general, I am fine if you say that it is not always what you would want to do).
Deleting both comments, but I am unsure that this actually clarifies things.

Reply:
@eric-wieser I am still struggling with what you write, in your gist:

> So we can support new dtypes defined using the class statement even if issubclass(np.dtype, type) is false.

Yes, IMO it should be `isinstance(np.dtype, type)`, now your `NotAType` seems a bit freaky? Since it has `not issubclass(type(NotAType), type)`. For one I do not see how it could possibly be initialized reasonably, since it will not have an `__init__()` and instead only has a `__call__()` since it is an object pretending to be a type. That is all fine, but it seems more confusing/limiting than helpful?

> dtype instantiation (not to be confused with invoking np.dtype itself) means just what it does today  [arr = uint8_non_native([1, 2, 3])]

This is not what it means today? That is the strange (I would prefer to deprecate) behaviour of *scalar* instanciation.

---

The only thing I currently managed to read out of your gist that you are thinking about putting *everything* on the same level as to remove the `class <-> instance` split I am proposing.
That is of course a possibility, but it seems weird for things like `metadata` (including "metadata" such as Unit information)? I do not think it is desirable to go through a subclassing (like) procedure/thinking to attach a different physical unit or string length. Also, I would like to be able to have some representation of `"S"` which is a string without a length. A string without a length is *not* a dtype in the sense that it cannot be attached to an array. For scalar-types, `np.integer` has no (direct) instances either. Representing it as the `DType` class (subclass of `np.dtype`) seems very elegant to me.
I do see that users are used to mainly work with dtype instances right now. And I do not expect that to change *most* datatypes. But I am unable to see the value in trying to flatten the distinction betwen DType classes defining the method/logic like a Python class and the dtype instance that is attached to an array. Yes, the (almost) singleton instances are a bit annoying, and the fact `np.dtype(np.int8)` can be mapped to `np.dtype("=int8")` while `np.dtype(np.string)` cannot be mapped (what is the strings length) *is* annoying. Since nobody will or should normally write `np.dtype[np.int8]()`...

Currently we have things like `.newbyteorder()` if you allow subclassing to add a "strange" byteorder, things will get tricky. In my model, every extension DType has to be its own class, there could sometimes be a shame (i.e. a UFunc would have to be told that it can use the `uint8` instead of the `uint8_strange_byteorder` loop. But I doubt it will be a big issue in practice and cleanly separates user datatypes and NumPy datatypes.

Reply:
> Yes, IMO it should be `isinstance(np.dtype, type)`

Agreed, but the the two are orthogonal so I thought I'd be clear

> now your NotAType seems a bit freaky? Since it has not issubclass(type(NotAType), type). For one I do not see how it could possibly be initialized reasonably, since it will not have an __init__() and instead only has a __call__() since it is an object pretending to be a type.

It's not an object pretending to be a `type`, it's an object _being_ a `dtype`. `dtype`s have no `__init__`, because they're created via arrays, not directly. `NotAType` doesn't have a `__call__` either, but `type(NotAType)` does.

Conventionally, `class` definitions are a means of filling C slots in `PyTypeObject`. That example shows `class` definitions being used to fill slots in `PyArray_Descr`. What I'm trying to show is that it doesn't matter whether the slots being filled in belong to `type` or `dtype`, `class` is still a useful syntactic construct, and base classes still mean slot defaults.



> This is not what it means today? That is the strange (I would prefer to deprecate) behaviour of scalar instanciation.

Ah, that is a good point. It's still an example of how that could be done, but agreed it's not super necessary.


> The only thing I currently managed to read out of your gist that you are thinking about putting everything on the same level as to remove the class <-> instance split I am proposing.

Yes, I think that's the core of my proposal, which is motivated by the way that python handles scalar types. An interesting exercise would be to try and apply your ufunc-like dispatch system for a set of scalar only types implemented in pure python - would you need meta-types? Meta-metatypes? 

> That is of course a possibility, but it seems weird for things like metadata (including "metadata" such as Unit information)? 

Perhaps, perhaps not. Under that design, `issubdtype(np.int8.with_metadata(...), np.int8)` is true but not the reverse, which is perhaps tidier than today's behavior. If you wanted to create a regular `list` type with metadata, you'd subclass `list`, so it's sort of consistent

Regarding units, I think I should reread the NEP and try and translate the proposal here to how it would look in the gist.

Reply:
(Maybe replying too fast) OK, maybe we can focus on the "flattening" then (no instance <-> class split). One thing I am not 100% sure about is what `isinstance(np.dtype("int8"), np.dtype)` gives in your thoughts, I assume it is or can be made to be True (that is important for backcompat)?

> dtypes have no __init__, because they're created via arrays, not directly.

Fair point, but `arr.dtype` has no reason not to return the initialized object? It still seems strange to me to hide that initialization step from the user.
I do not disagree with the technical fact that it could be fine to not inherit from `type`, but doing it at least gives us useful `issinstance` and `issubclass` definitions?

> It's still an example of how that could be done, but agreed it's not super necessary. [About `int8([1, 2, 3, 4])` creating an array]

Yes, but that conflates the array object and the dtype, and I would like to keep them as separated as possible. That may end up being a fairly useless attempt, but it is a principle we should not give up quite yet?

> `issubdtype(np.int8.with_metadata(...), np.int8)` is true but not the reverse

Hmmm, I still feel `issubdtype(">i8", "<i8")` is a red herring in that it is a weird question to ask. In your thoughts it seems largely the same as `issubclass()`? But that cannot be commutative. In my proposal it would be deprecated and `isinstance(">i8", type(np.dtype("i8")))` would be preferred (np.dtype[np.int64]` and just `np.int64` are both acceptable alternative spellings for `type(np.dtype("i8"))`).

`issubdtype` supports the abstract scalar class hierarchy. For typical dtypes it is the same as comparing `type_num`, though, and I assume that `type_num` is a necessary concept – although maybe not in `issubdtype`.

This leads to a break in the hierarchy. `>int8` and `<int8` do not have an "order", they seemat same level but distinct e.g. from `<int64`. You could have an `all_int8s` class with  `<int8` and `>int8` as subclasses, but `<int8` and `>int8` are leaves of the hierarchy tree. And it seems to me that whether scalar or not, instances are the normal way to represent the leaves in such a tree.

I have to admit though, I am not sure what you are actually have in mind for `issubdtype(">i8", "<i8")` (error, True, or False).

> An interesting exercise would be to try and apply your ufunc-like dispatch system for a set of scalar only types implemented in pure python - would you need meta-types? Meta-metatypes?

I plan to propose multiple dispatching based on the `type(arr.dtype)`, with some limitations since I do not want to allow arbitrary subclassing for concrete DTypes (at the very least initially and w.r.t. inheritance). Just like a scalar multiple dispatching would use `type(scalar)`. So there is simply no difference IMO.

The problem is after that "dispatching" step you still could have different byte-orders, etc. so that casting may be necessary (ignoring fallback to loops that do not match exactly, by using abstract superclasses). There are no MetaClasses or MetaDTypes involved, but after dispatching the casting requires some of the `np.dtype` base-classes slots (dunders) – and yes, those require a MetaClass to look them up very quickly on the C-side.

> Regarding units, I think I should reread the NEP and try and translate the proposal here to how it would look in the gist.

I think even string may be enough. Because right now I do not see where the `string-of-given-length` factory would go naturally, in this proposal it is `np.dtype[np.string](length=10)` (where `np.dtype[np.string]` could have its own top-level name of course).
I am trying to abstract the "flexible" part away. "flexible dtypes" are classes, dtype instances cannot be "flexible".

Reply:
So I am trying to get back to this... and what the differences between my proposal here and your thoughts are.

The more I look at it, the less I see the relevant user differences to be honest, I have probably written similar stuff above.

You are proposing an "`issubdtype`" logic, when for me:
```
def issubdtype(dtype1, dtype2):
    assert isinstance(dtype2, np.dtype)
    isinstance(dtype1, type(dtype2))
```
(Although I have no desire to actually use subclasses for now, it keeps things simpler). We can probably agree that this hierarchy is necessary, e.g. for dispatching in ufuncs and casting. Whether `uint8_with_metadata is below uint8` matters little. 

My proposal has the downside that we have to use MetaClasses much more, and most relevant data is stored on those MetaClasses. That makes the dtype instances a bit thinner in theory, although that does not matter much. In practice, if you would want to have specialized fast-paths for some `>int8` functions, my approach is less nice. You could have an `>int8` subclass as well (making it basically true singleton instances), but that would require a fat type object.
To be honest: I do not think MetaClasses as an implementation detail matter, they are invisible for the user, fat objects do not matter for dtypes (unless you have hundrets of them).

So what are the actual differences?

1. Inheritance: `>int64` cannot actually inherit everything from from `<int64`. E.g. a comparison function is not byte-compatible. `int64` could inherit certain things from `integer` maybe. I doubt we want users to really use inheritance much in either case though.

2. `scalar types` ⇆ `dtypes or DTypes`. There are two sides to this "correspondence":
   a) In my proposal there is a direct mapping from a scalar type to the corresponding `DType`. This `DType` knows details such as storage and can be inspected by looking at its initializer. In particular the `Dtype` corresponds to an actual type and is not just a class (a set of values which can be represented).
   b) In your thoughts, you have a strong correspondence when it comes to methods. The `dtype` acts like a `type`, it is passed in the "instance", although here the instance is a pointer into the array.
  
  The main problem I see with that approach is how strings fit in. At least when it comes to strings, you fully lose the mapping property in a) in your proposal. Also you never have the type correspondence, you have the class-purpose correspondence in b).
  But I am not sure they fully correspond to a type in b) since the string length has to be passed in as well? I.e. you still need to pass in the `dtype` (as opposed to just its value), even though the method already lives on the dtype. That is not bad, but I think that the user-side correspondence in a) is more important? And passing in the `instance` as a `self` argument is a nice correspondence as well, and it is one that I have.
  In other words: I say that `scalar instance` ⇆ `array instance + DTypes instance`, which seems pretty clear about the fact that a method living on the `DType` needs both information from the array (the element) and the dtype?

So what are the actual downsides of the proposal in this NEP?

* I suppose I cannot optimize as easily for certain string lengths or specialize for byte-order on certain levels. This is semi intentionally, since we currently do not do it, but it is a possible downside.
* I have to wrangle with MetaClasses, I do not see this as a major issue though, mainly an annoyance.
* We have a lot of DTypes which have, for all practical purposes, singleton instances. And often we will have to allow you to pass in the DType and silently create the instance from it for you.
* I need abstract DTypes, and additional metaclasses to handle value based casting probably, that sucks... But, value based casting is a tricky beast, so, yes you could avoid having that require a MetaClass, but in the end the hacks will be similar, just housed in a slightly more lean object.

The upsides in my opinion are:

* We have a clear type hierarchy implemented the same way as the scalar type hierarchy and with classes. And the DType class/type corresponds to the actual scalar type (as in which values can be represented).
* The step of having `dtype` instances means parametric `DType`s have a clear API, `np.dtype[np.unicode]` can return something useful which can be *instantiated* to something useful.
* We can get rid of `issubdtype` or implement it directly based on the Python class hierarchy.

What Travis was thinking of, to get rid of the `DType` class would be to put everything on the scalar. Maybe what he was thinking of is that you would have a special method `np.int64.get_dtype(byteorder="=")`, which has its merit. In that view `type(np.int64.get_dtype(byteorder="="))` would be reasonable as your `integer` (although just `dtype` would be just as well within NumPy), and we might indeed store all dtype functionality on the instance instead of the class.
For Python builtin types we cannot do this of course, so this would mean we need some "dummy" scalars (and map to them).

But as much as I always like that, I somewhat fail to see that it will actually help much. We still need to define a bunch of structs (namely the current `->f`, etc.) which _effectively_ are much like defining a full scalar python class/type. A `np.int64.get_dtype()` method is neat, but I am not sure that returning an instance of a `np.dtype[np.int64]` is actually bad, at least it allows you to keep your scalars fully free of NumPy.

Reply:
OK, so to continue on that. If we say dtypes are simply a "generic" object describing the data, without additional bells and whistles, then I think the scalars have to take the brunt of it:

1. We will have to add at least some slots to the scalar:
    * Possibly just be the class method `__new_dtype__(*args, **kwargs)` and a method `__dtype__()`.
    * Casting needs a home somewhere, although it could be implemented more like ufuncs as multimethods if scalars were not MetaClasses.
2. This means that:
    * Scalars might need to be MetaClass instances, or `->f` slots need to be tacked on to every dtype instance (`->f` slots are much like a "class" IMO, but it may actually be neater/easier to avoid reference cycles between scalar types and DTypes).
    * Existing python scalars cannot add any additional methods. So we will need fake scalars, similar to `np.object_`, I think.
    * DType classes become an implementation detail. Subclassing `np.dtype` is only necessary for extension types that wish to add methods or store units, etc.. `np.isdtype` or similar exists, but you might as well work with `dtype.type` probably.
    * The scalar class hierarchy would be important if we want to use a hierarchy for dispatching! (which I currently see it being used for). Even if not, it would be the scalars providing the hierarchy in some form.
    * The DType classes would probably be considered to have less features, which may be good. The downside is if we have "fake" scalars like `np.object_`. If we have DType classes next to scalar classes, `np.dtype[scalar_type]` can always give you what you want (I admit doing the mapping is annoying). If someone wants to store `datetime.datetime` directly we need the same mapping somewhere, but there seems no "obvious" way to expose it to the user, which means there is no obvious way to do `datetime.datetime.get_dtype()` or similar. OTOH, I suppose you can still do a function approach. `np.dtype(datetime.datetime)` can still work, although it seems a bit tedious if the `datetime.datetime` DType were supposed to be parametric (where do you read up on the arguments/kwargs?).

 So... Maybe we can avoid metaclasses (as in special storage) by pushing things more into the dtype instances (more like it is now). A tricky part seems to me that we have to "identify" scalars that have DTypes, duck typing might work but we may need to use an ABC or even a MetaClass, which I am not sure I like to force up on scalars.

The biggest issue I still have is that it seems to me pushing things into the scalar seems tricky if the scalar is supposed to know nothing about NumPy. I.e. it would almost need to be a language feature in Python?
One of the larger annoyances I have with my metaclasses, is that the relation `Scalar Type<->DType type` may be a bit annoying. In some cases the scalar need not know anything about the `DType`, but in practice at least some of our scalars must. And I doubt that circular references even matter enough, the GC tracks tens of thousands of objects on a fresh IPython start...

Anyway, babbling by now :). I am not getting closer to being absolutely sure that one option is better than the other. And I am not sure that this actually has too much to do with your thoughts Eric.

In the end maybe my mistake is to think that normal Python scalars can/should have dtypes, and we should bake in that any scalar that has a "dtype" (can be stored directly, including discovery) to be derived from an `np.scalar`...

Reply:
Hopefully I'll be able to come back to this conversation in the next week, just dropping in to assure you your thoughts are not going directly into the void.

Reply:
I find it hard to pin down the actual up and downsides. And of course your thoughts are probably in different direction anyway. Since I disliking the idea of a "hierarchy" outside of the class hierarchy. Of course there are arguments for using a "safe-casting" hierarchy, but I also see difficulties with that (although it sometimes generalizes a bit neater).

Up sides to a scalar-centric view:
  * Probably less MetaClassing
  * Less danger of circular referencing (although I doubt it matters)
  * No twin-implementation DType <-> Scalar type
  * No twin, possibly differing, hierarchies.
  * No twin implementation means no/less danger of random people implementing a DType for a random scalar.

Down sides:
  * Restricting inheritance/usage may be less desirable for scalars than for DType classes.
  * The scalar type must implement a certain interface (i.e. inheriting from `np.generic`, maybe even be a MetaClass instance). Or at least implement certain dunders. 
  * Thus: Difficult or impossible to implement DType for an arbitrary scalar type.
  * arrays rely on dispatching (unlike python scalars). The hierarchy is thus more important and allows for more confusion than typical Python scalars. Whatever hierarchy we use for ufunc dispatching may need to be cleaner than a typical Python scalar hierarchy.

User Impact:
  * DTypes such as `categorical` or `object` where the element is an arbitrary python object are more weird if the scalar type is central. But: We may actually say that `object` is the _only_ exception and a `categorical` result should be like an `enum.Enum` instance. A similar thing could be for an array of `np.ndarray`.
  * API Can either look more like:
    * `scalar_type.get_dtype(*args, **kwargs)`
    * `np.dtype[scalar_type](*args, **kwargs)`
    * If we allow dtypes without a clear scalar type, both have to use: `something(*args, **kwargs)`, in the scalar-centric case, I am not sure I have a good story, in the second you still need a `DType` class though. (The dtype type can be used also for the scalar-centric story, but I feel it has not quite fit in anywhere)

Implementation Impact:
* If the user has no scalar:
  * Scalar-centric: Create a new scalar with a NumPy provided API, or create a scalar and add a certain protocol on. They may need to create a `np.dtype` subclass as well.
  * DType-centric: Create a DType class and we have to help them to create a scalar as well.
* If the user already has a scalar:
  * Scalar-centric: The user has to extend the scalar and create a dtype subclass if necessary.
  * DType-centric: A DType is created, the scalar can remain unmodified (at least in principle)

---

But I guess, you may be coming from a bit different direction. And that is that the methods are actually more naturally defined on what we call `dtype` now, since it knows the exact data representation "type" of each element. And from that direction, I don't have much to add, except maybe
```
class Int8ScalarWithStorageDetails():
    byteorder = "="  # part of dtype
    byte_data 
```
also would store the byteorder on the instance.

Reply:
After a long time, in an attempt at resolving this discussion, or make my point of why this decision. I started to think about [here](https://hackmd.io/m6rhLhAqTmuhObW62N4sQA) (yes, its too long).

I will maintain what I think are my old points:

1. We need some representation of a "String" (i.e. `np.dtype("S")`).  There should be a clear place where value space and behaviour is *defined* (but not necessarily implemented).
   * `type(np.dtype("S5"))` can live up to that task
   * We have a clear relation that the value space and behaviour of one scalar-type is fully described by one DType 
   * In the model where the `dtype` instance is considered the implementer of behaviour, we need an
     abstract base class for each dtype: `np.dtype("S")` is *almost* that.
     We also need either that ABC or the scalar type to take the role of defining behaviour.
     (The scalar type is great, but I prefer it knows as little as possible about arrays and dtypes.)

2. Parametric `dtypes` are also normal python instances. There is some behaviour such as `np.promote_types()` and the NEP 42 `DType.__discover_descriptor__(python_object)` which has nothing to do with array values, and so it is clearly the `DType` class (meta-dtype) which should provide this behaviour.

At this point I said that I want something which "defines" (but not necessary implements) the value space and behaviour of a dtype.  And I claim, that the DType class (meta-dtype) is the natural choice, because we need it anyway and because otherwise the Python model would be to have an "abstract base dtype" (equivalent to our "flexible"  `np.dtype("S")` which technically shouldn't be attached to an array) instead.

So my perception is: We need a DType class and that DType class will *effectively* define/represent value space and behaviour.  Even in your model, where array element behaviour should be implemented on the `dtype` we have to deal with this DType class (it *might* be possible to make it more generic), but it must exist.

However: It does not need to *implement* the behaviour of the array elements.  We could look up behaviour just as well on the dtype directly.  Our current C code looks like that in places, since it uses:
```
PyArray_Descr *descr = PyArray_DESCR(arr);
descr->f->setitem(...)
```
Note that there is a *big* but here: `descr->f` is *always* shared for all dtypes with the same type number.  There is never any inheritance, so that the above code could just as well be:
```
PyArray_Descr *descr = PyArray_DESCR(arr);
descr->tp_type->setitem(...)
```
in all cases. And NEP 41 to, to some degree, requires this (see below to resolve this).

<details> <summary> In NEP 41, we consider the "instance" of a single array value to be `dtype+element`: </summary>

If we look inside an Python float, it has the structure:
```
float instance:
    type = float  # The instance stores its type
    value = 4.
```
The `dtype` itself is a Python type (spelling `type(np.dtype("float64"))` as `Float64`):
```
Float64 instance:
    type = Float64
    byte_order = "<"
```
While the individual array value can be thought of:
```
Float64 value:
    dtype = np.dtype("<float64")  # Stored on the array
    value = 4.
```
NEP 41, basically extends the value "instance"  rather than defining the type:
```
Float64 value:
    type = Float64       # stored in `arr.dtype`
    byte_order = "<"    # stored in `arr.dtype`
    value = 4.
```

</details>

---

The issue that I think we never talked about is the *why* you would want to store/implement behaviour on the `dtype` rather than its type.  And the main reason I can think of is to "specialize" functionality for speed.

My answer to the need of specialization is to think of methods similar to unbound methods in Python, i.e. this is what ufuncs do (or will do):

<details>

**Kernels (DType Methods) resolve the lack of specificity:**

A Python method is itself a class with a special `__get__` method. When the `.` operator fetches a method from a class, it binds it by calling `__get__`. We can write our own "method" class:
```
from functools import partial

def func(*args, **kwargs):
    print(args, kwargs)

class Method:
    _func = func  # function written by user

    def __get__(self, instance, owner=None):
        return partial(self._func, instance)

# Note that `__get__` is called and Method works:
class MyObject:
    meth = Method

bound_method = MyObject().meth  # calls __get__
bound_method()  # prints self
```

We can (and in casting to some extend do) translate this to
saying that a "kernel" (DType method), also can have
a `__get__` method:

```
class Kernel:
    _func1 = function_for_little_endian
    _func2 = function_for_big_endian
    
    def __get__(self, dtype):
        if dtype.byteorder == "<":
            return _func1
        else:
            return _func2
```
Making it possible to write functions specific to a a certain instance
(just as it would be possible in pure Python even *without* inheritance).
Although, the *organization* of these specific function occurs
one level higher.

This `__get__` method doesn't need to be exposed right away, but it is a clear
extension if "kernels" define at least almost all functionality of array values.
(Which is the clear goal with respect to ufuncs and casting.)

Once we expose the `__get__` function, the DType implementer could even create their own system of inheritance by looking up behaviour on the dtype instance itself (I guess similar to monkey patching a Python instance?).

</details>

Reply:
I hope to find time to read this in the next month or so...